contributors/design-proposals/grow-volume-size.md

```diff
@@ -0,0 +1,253 @@
+new controller will be:
+
+* Watch for pvc update requests and add pvc to controller's desired state of world if a increase in volume size was requested. Once PVC is added to
+  controller's desired state of world - `pvc.Status.Conditions` will be updated with `ResizeStarted: True`.
+* For unbound or pending PVCs - resize will trigger no action in `volume_expand_controller`.
+* A reconciler will read desired state of world and perform corresponding volume resize operation. If there is a resize operation in progress
+  for same volume then resize request will be pending and retried once previous resize request has completed.
+* Controller resize in effect will be level based rather than edge based. If there are more than one pending resize request for same PVC then
+  new resize requests for same PVC will replace older pending request.
+* Resize will be performed via volume plugin interface, executed inside a goroutine spawned by `operation_exectutor`.
+* A new plugin interface called `volume.Expander` will be added to volume plugin interface. The controller call to expand the PVC will look like:
+
+```go
+func (og *operationGenerator) GenerateExpandVolumeFunc(
+    pvcWithResizeRequest *expandcache.PvcWithResizeRequest,
+    dsow expandcache.DesiredStateOfWorld) (func() error, error) {
+
+    volumePlugin, err := og.volumePluginMgr.FindExpandablePluginBySpec(pvcWithResizeRequest.VolumeSpec)
+
+    if err != nil {
+        return nil, fmt.Errorf("Error finding plugin for expanding volume: %q with error %v", pvcWithResizeRequest.UniquePvcKey(), err)
+    }
+
+    expanderPlugin, err := volumePlugin.NewExpander()
+
+    if err != nil {
+        return nil, fmt.Errorf("Error creating expander plugin for volume %q with error %v", pvcWithResizeRequest.UniquePvcKey(), err)
+    }
+
+    expandFunc := func() error {
+        expandErr := expanderPlugin.ExpandVolumeDevice(pvcWithResizeRequest.VolumeSpec, pvcWithResizeRequest.ExpectedSize, pvcWithResizeRequest.CurrentSize)
+
+        if expandErr != nil {
+            glog.Errorf("Error expanding volume through cloudprovider : %v", expandErr)
+            return expandErr
+        }
+        dsow.MarkAsResized(pvcWithResizeRequest)
+
+        return nil
+    }
+    return expandFunc, nil
+}
+```
+
+* Once volume expand is successful, the volume will be marked as expanded and new size will be updated in `pv.spec.capacity`. Any errors will be reported as *events* on PVC object.
+* If resize failed in above step, in addition to events - `pvc.Status.Conditions` will be updated with `ResizeFailed: True`. Corresponding error will be added to condition field as well.
+* Depending on volume type next steps would be:
+
+    * If volume is of type that does not require file system resize, then `pvc.status.capacity` will be immediately updated to reflect new size. This would conclude the volume expand operation. Also `pvc.Status.Conditions` will be updated with `Ready: True`.
```

Comment:
If admin manually modifies `pv.spec.capacity` - I am thinking, we won't do any cloudprovider call to resize if the new size meets the critirea of PVCs. i.e - before considering a volume for resize in controller-manager, we will also compare pv.spec.capacity with requested size. It may be that, pv already satisfies the requested size and no cloudprovider based call to resize is necessary. 

This isn't captured in proposal, but I will add it. hopefully, that covers potential race you are talking abou.t